### PR TITLE
Fix incremental bugs in the HIR map

### PR DESCRIPTION
**Pull request by Zoxc**

### CODE REVIEW

**Comment by Zoxc** [comment]
I added a `parent_module` query and that seems to have fixed the large regressions. `unused-warnings` regresses a bit due to the additional `parent_module` query.

r? @michaelwoerister

**Comment by michaelwoerister** [comment]
Are these perf results still current?

**Comment by Zoxc** [comment]
Yeah. I just removed a commit and it's reversion.

File: src/librustc/hir/map/mod.rs [hunk 1]
```diff
@@ -612,7 +618,12 @@ impl<'hir> Map<'hir> {
         if self.dep_graph.is_fully_enabled() {
             let hir_id_owner = hir_id.owner;
             let def_path_hash = self.definitions.def_path_hash(hir_id_owner);
-            self.dep_graph.read(def_path_hash.to_dep_node(DepKind::HirBody));
+            let kind = if hir_id.local_id == ItemLocalId::from_u32_const(0) {
+                DepKind::Hir
+            } else {
+                DepKind::HirBody
+            };
+            self.dep_graph.read(def_path_hash.to_dep_node(kind));
```
Comment (michaelwoerister) [inline src/librustc/hir/map/mod.rs]: I don't understand what this logic is supposed to do exactly (before or after the change)...

Comment (Zoxc) [reply src/librustc/hir/map/mod.rs]: It is supposed to add a `read` to the node which reveals the parent of the `HirId`.

Comment (michaelwoerister) [reply src/librustc/hir/map/mod.rs]: OK, I understand the original thinking now: The fingerprint of `HirBody` contains the information of the entire item (including the "signature" parts) so adding a dependency to `HirBody` is the conservative choice. 

But as you discovered the `parent` is not hashed into any of the two dep-nodes, so we are clearly missing dependencies here.

I assume that the `if hir_id.local_id == ItemLocalId::from_u32_const(0)` check is based on the assumption that the parent of any node except the item root is part of the same HIR item and thus the `HirBody` fingerprint contains the appropriate information? If that's the case it is far from obvious and a clarifying comment will be much appreciated by anybody trying to understand this code.

File: src/librustc/hir/mod.rs [hunk 2]
```diff
@@ -46,9 +47,17 @@ impl<'tcx> TyCtxt<'tcx> {
     pub fn hir(self) -> Hir<'tcx> {
         Hir { tcx: self, map: &self.hir_map }
     }
+
+    pub fn hir_id_parent_module(self, id: HirId) -> DefId {
```
Comment (michaelwoerister) [inline src/librustc/hir/mod.rs]: Maybe `parent_hir_module` would be more appropriate.

File: src/librustc/hir/mod.rs [hunk 3]
```diff
@@ -46,9 +47,17 @@ impl<'tcx> TyCtxt<'tcx> {
...
+        self.parent_module(DefId::local(id.owner))
+    }
 }
 
 pub fn provide(providers: &mut Providers<'_>) {
+    providers.parent_module = |tcx, id| {
+        let hir = tcx.hir();
+        hir.get_module_parent(hir.as_local_hir_id(id).unwrap())
```
Comment (michaelwoerister) [inline src/librustc/hir/mod.rs]: Can we make this function `get_module_parent` and `get_module_parent_node` less public? The query is always the better choice, right?

**Comment by michaelwoerister** [comment]
The fix looks good to me. Am I correct in assuming the https://github.com/rust-lang/rust/pull/68944 will make the whole thing obsolete?

**Comment by Zoxc** [comment]
@michaelwoerister https://github.com/rust-lang/rust/pull/68944 includes the `parent_module` commit, but otherwise yes.

File: src/librustc/hir/mod.rs (hunk 2 above)
Comment (Zoxc) [reply src/librustc/hir/mod.rs]: Maybe renaming the query to `parent_module_from_def_id` and letting this take `parent_module` is a better option?

**Pull request closed without merging**